### PR TITLE
Resolve #1169 -- Bush Logic

### LINKED ISSUE #1169: Bush Logic
Add vision checks to handle bushes using ray casting.

### PR DESCRIPTION
**Pull request by cabeca1143**
* Vision checks now handle bushes.

Resolve #1169

### CODE REVIEW

**Review by danil179 — APPROVED** [review]
Seems like it should work, I'll test it in the next few days. I think that you can optimize some of the checks by saving the parameters (as HasFlag(destination, NavigationGridCellFlags.HAS_GRASS) or HasFlag(origin, NavigationGridCellFlags.HAS_GRASS) doesn't change). This ray tracing function is using a lot of CPU and getting called many times, so it is important to check the performance.

**Comment by cabeca1143** [comment]
> Seems like it should work, I'll test it in the next few days. I think that you can optimize some of the checks by saving the parameters (as HasFlag(destination, NavigationGridCellFlags.HAS_GRASS) or HasFlag(origin, NavigationGridCellFlags.HAS_GRASS) doesn't change). This ray tracing function is using a lot of CPU and getting called many times, so it is important to check the performance.

Addressed

File: GameServerLib/Content/Navigation/NavigationGrid.cs [hunk 1]
```diff
@@ -803,15 +803,37 @@ public float GetHeightAtLocation(float x, float y)
...
+                if (checkVisible && !prevPosHadBush)
+                {
+                    if (isGrass)
+                    {
+                        break;
+                    }
+                }
+
+                // If you are in a different bush
+                if (checkVisible && prevPosHadBush && HasFlag(destination, NavigationGridCellFlags.HAS_GRASS))
```
Comment (danil179) [inline GameServerLib/Content/Navigation/NavigationGrid.cs]: `HasFlag(destination, NavigationGridCellFlags.HAS_GRASS)` can also be computed outside of the loop I think, as the destination doesn't change.

**Review by danil179 — COMMENTED** [review]
Another possible optimization (not sure if the compiler does that already).

File: GameServerLib/Content/Navigation/NavigationGrid.cs (hunk 1 above)
Comment (cabeca1143) [reply GameServerLib/Content/Navigation/NavigationGrid.cs]: I could swear i had made that change o_0
Well here it is again(?)

**Review by danil179 — APPROVED** [review]
Approved.

**Pull request merged**